### PR TITLE
Type declaration bounds in inference

### PR DESCRIPTION
**Pull request by d367wang**
1. returns expected `@VarAnnot(?)` for `InferenceAnnotatedTypeFactory.getTypeDeclarationBounds`.
2. type declaration bound should be a normal source variable slot, instead of existential slot.

3. incorporate part of https://github.com/opprop/checker-framework-inference/pull/294, such that not to create existential slots for class bounds, i.e. type declaration bounds are always explicit in inference.

### CODE REVIEW

File: src/checkers/inference/VariableAnnotator.java [hunk 1]
```diff
@@ -1612,17 +1612,17 @@ protected Boolean scan(AnnotatedTypeMirror type, Void aVoid) {
...
 
         VariableSlot topConstant = getTopConstant();
         VariableSlot declSlot = classDeclAnnos.get(classDecl);
         if (declSlot == null) {
             Tree decl = inferenceTypeFactory.declarationFromElement(classDecl);
             if (decl != null) {
-                VariableSlot potentialDeclSlot = createVariable(decl);
-                declSlot = getOrCreateExistentialVariable(potentialDeclSlot, topConstant);
-                classDeclAnnos.put(classDecl, potentialDeclSlot);
+                // Since each class decl should have a slot,
```
Comment (d367wang) [inline src/checkers/inference/VariableAnnotator.java:1622]: remove

File: src/checkers/inference/VariableAnnotator.java [hunk 2]
```diff
@@ -1632,6 +1632,21 @@ private VariableSlot getOrCreateDeclBound(AnnotatedDeclaredType type) {
...
 
+    /**
+     * Get the {@link VarAnnot} on the class declaration of the Element.
+     * @param ele an element
+     * @return the {@link VarAnnot} on the class declaration,
+     * or {@code null} if the class declaration of the Element is not handled by the
+     * {@link VariableAnnotator#getOrCreateDeclBound(AnnotatedDeclaredType)}.
+     */
+    public AnnotationMirror getClassDeclVarAnnot(TypeElement ele) {
+        if (classDeclAnnos.get(ele) != null) {
```
Comment (d367wang) [inline src/checkers/inference/VariableAnnotator.java:1643]: anno = classDeclAnnos.get(ele)

File: src/checkers/inference/InferenceAnnotatedTypeFactory.java [hunk 3]
```diff
@@ -569,12 +574,24 @@ protected InferenceViewpointAdapter createViewpointAdapter() {
     }
 
     /**
-     * Get the real top annotation from {@link #realTypeFactory}.
-     * @return the real top annotation.
+     * Get the annotation from the class declaration.
+     * @param type a type
+     * @return the set of {@link VarAnnot} on the class bound. Unlike type checking,
```
Comment (d367wang) [inline src/checkers/inference/InferenceAnnotatedTypeFactory.java:579]: singleton set with VarAnnot on the class bound

File: src/checkers/inference/InferenceAnnotatedTypeFactory.java [hunk 4]
```diff
@@ -569,12 +574,24 @@ protected InferenceViewpointAdapter createViewpointAdapter() {
...
+     * the returned set is always singleton in inference.
+     * {@link org.checkerframework.framework.type.AnnotatedTypeFactory#getTypeDeclarationBounds}
+     * if the class is not handled by the {@link VariableAnnotator}.
      */
     @Override
-    protected Set<? extends AnnotationMirror> getDefaultTypeDeclarationBounds() {
-        return realTypeFactory.getQualifierHierarchy().getTopAnnotations();
+    public Set<AnnotationMirror> getTypeDeclarationBounds(TypeMirror type) {
+        final TypeElement elt = (TypeElement) getProcessingEnv().getTypeUtils().asElement(type);
+        AnnotationMirror vAnno =
```
Comment (d367wang) [inline src/checkers/inference/InferenceAnnotatedTypeFactory.java:587]: into one line

File: src/checkers/inference/VariableAnnotator.java [hunk 5]
```diff
@@ -1632,6 +1630,22 @@ private VariableSlot getOrCreateDeclBound(AnnotatedDeclaredType type) {
         return declSlot;
     }
 
+    /**
+     * Get the {@link VarAnnot} on the class declaration of the Element.
+     * @param ele an element
```
Comment (d367wang) [inline src/checkers/inference/VariableAnnotator.java:1635]: type element

**Review by wmdietl — APPROVED** [review]

**Pull request merged**